### PR TITLE
Remove edition umbrella features.

### PR DESCRIPTION
**Pull request by ehuss**
In the 2018 edition, there was an "umbrella" feature `#[feature(rust_2018_preview)]` which was used to enable several other features at once. This umbrella mechanism was not used in the 2021 edition and likely will not be used in 2024 either. During 2018 users reported that setting the feature was awkward, especially since they already needed to opt-in via the edition mechanism.

This PR removes this mechanism because I believe it will not be used (and will clean up and simplify the code). I believe that there are better ways to handle features and editions. In short:

- For highly experimental features, that may or may not be involved in an edition, they can implement regular feature gates like `tcx.features().my_feature`.
- For experimental features that *might* be involved in an edition, they should implement gates with `tcx.features().my_feature && span.at_least_rust_20xx()`. This requires the user to still specify `#![feature(my_feature)]`, to avoid disrupting testing of other edition features which are ready and have been accepted within the edition.
- For experimental features that have graduated to definitely be part of an edition, they should implement gates with `tcx.features().my_feature || span.at_least_rust_20xx()`, or just remove the feature check altogether and just check `span.at_least_rust_20xx()`.
- For relatively simple changes, they can skip the whole feature gating thing and just check `span.at_least_rust_20xx()`, and rely on the instability of the edition itself (which requires `-Zunstable-options`) to gate it.

I am working on documenting all of this in the rustc-dev-guide.

### CODE REVIEW

File: compiler/rustc_error_codes/src/error_codes.rs [hunk 1]
```diff
@@ -429,7 +429,6 @@ E0700: include_str!("./error_codes/E0700.md"),
 E0701: include_str!("./error_codes/E0701.md"),
 E0703: include_str!("./error_codes/E0703.md"),
 E0704: include_str!("./error_codes/E0704.md"),
-E0705: include_str!("./error_codes/E0705.md"),
```
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes.rs:432]: Please put it back and update the explanation to say it's not emitted anymore (CI will fail in any case explaining this too :wink:).

Comment (ehuss) [reply compiler/rustc_error_codes/src/error_codes.rs]: Yep. I saw others were "removed" and didn't bother to read all the comments.

File: compiler/rustc_error_codes/src/error_codes/E0705.md [hunk 2]
```diff
@@ -1,3 +1,5 @@
+#### Note: this error code is no longer emitted by the compiler.
```
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes/E0705.md:1]: You'll need to remove `E0705` from the code blocks labels below as well.

Comment (ehuss) [reply compiler/rustc_error_codes/src/error_codes/E0705.md:1]: OK, I was a little confused why that wasn't generating an error. The `rust2018` tag was incorrect, and was causing rustdoc to ignore the code block.

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0705.md:1]: Oh interesting. Might be interesting to add a check for that then. Adding that to my todo list.

**Comment by hi-rustin** [comment]
Note: This PR has broken the cargo nightly tests. See: https://github.com/rust-lang/cargo/pull/13152#issuecomment-1851206780

**Comment by nnethercote** [comment]
Note: [this section](https://rustc-dev-guide.rust-lang.org/implementing_new_features.html#stability-in-code) of the rustc-dev-guide needs updating, to remove the bits about editions.

**Comment by ehuss** [comment]
Thanks for the heads up! Posted the fix at https://github.com/rust-lang/rustc-dev-guide/pull/1836.

**Pull request merged**